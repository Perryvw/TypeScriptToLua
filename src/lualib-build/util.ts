import * as tstl from "..";

export function isExportTableDeclaration(node: tstl.Node): node is tstl.VariableDeclarationStatement & { left: [] } {
    return tstl.isVariableDeclarationStatement(node) && isExportTable(node.left[0]);
}

export function isExportTable(node: tstl.Node): node is tstl.Identifier {
    return tstl.isIdentifier(node) && node.text === "____exports";
}

export type ExportTableIndex = tstl.TableIndexExpression & { index: tstl.StringLiteral };
export function isExportTableIndex(node: tstl.Node): node is ExportTableIndex {
    return tstl.isTableIndexExpression(node) && isExportTable(node.table) && tstl.isStringLiteral(node.index);
}

export function isExportAlias(
    node: tstl.Node,
): node is tstl.VariableDeclarationStatement & { right: [ExportTableIndex] } {
    return tstl.isVariableDeclarationStatement(node) && node.right !== undefined && isExportTableIndex(node.right[0]);
}

export type ExportAssignment = tstl.AssignmentStatement & { left: [ExportTableIndex] };
export function isExportAssignment(node: tstl.Node): node is ExportAssignment {
    return tstl.isAssignmentStatement(node) && isExportTableIndex(node.left[0]);
}

export function isRequire(node: tstl.Node) {
    return (
        tstl.isVariableDeclarationStatement(node) &&
        node.right &&
        tstl.isCallExpression(node.right[0]) &&
        tstl.isIdentifier(node.right[0].expression) &&
        node.right[0].expression.text === "require"
    );
}

export function isImport(node: tstl.Node, importNames: Set<string>) {
    return tstl.isVariableDeclarationStatement(node) && importNames.has(node.left[0].text);
}

export function isExportsReturn(node: tstl.Node) {
    return (
        tstl.isReturnStatement(node) &&
        tstl.isIdentifier(node.expressions[0]) &&
        node.expressions[0].text === "____exports"
    );
}
