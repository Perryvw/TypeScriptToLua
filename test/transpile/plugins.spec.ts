import * as util from "../util";
import { resolveFixture } from "./run";

test("printer", () => {
    util.testModule``
        .setOptions({ luaPlugins: [{ name: resolveFixture("plugins/printer.ts") }] })
        .expectLuaToMatchSnapshot();
});

test("visitor", () => {
    util.testFunction`
        return false;
    `
        .setOptions({ luaPlugins: [{ name: resolveFixture("plugins/visitor.ts") }] })
        .expectToEqual(true);
});

test("visitor using super", () => {
    util.testFunction`
        return "foo";
    `
        .setOptions({ luaPlugins: [{ name: resolveFixture("plugins/visitor-super.ts") }] })
        .expectToEqual("bar");
});

test("getModuleId", () => {
    util.testBundle`
        export { value } from "./foo";
    `
        .addExtraFile("foo.ts", "export const value = true;")
        .setOptions({ luaPlugins: [{ name: resolveFixture("plugins/getModuleId.ts") }] })
        .expectToEqual({ value: true })
        .expectLuaToMatchSnapshot();
});

test("getResolvePlugins", () => {
    util.testBundle`
        export { value } from "foo";
    `
        .addExtraFile("bar.ts", "export const value = true;")
        .setOptions({
            luaPlugins: [{ name: resolveFixture("plugins/getResolvePlugins.ts") }],
            baseUrl: ".",
            paths: { foo: ["bar"] },
        })
        .expectToEqual({ value: true });
});

test("passing arguments", () => {
    util.testFunction`
        return {};
    `
        .setOptions({ luaPlugins: [{ name: resolveFixture("plugins/arguments.ts"), option: true }] })
        .expectToEqual({ name: resolveFixture("plugins/arguments.ts"), option: true });
});
