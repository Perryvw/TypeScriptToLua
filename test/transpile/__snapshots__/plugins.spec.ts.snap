// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getModuleId 1`] = `
"
local ____modules = {}
local ____moduleCache = {}
local ____originalRequire = require
local function require(file)
    if ____moduleCache[file] then
        return ____moduleCache[file]
    end
    if ____modules[file] then
        ____moduleCache[file] = ____modules[file]()
        return ____moduleCache[file]
    else
        if ____originalRequire then
            return ____originalRequire(file)
        else
            error(\\"module '\\" .. file .. \\"' not found\\")
        end
    end
end
____modules = {
[\\"5d05566c99dac259f6ff5742a268d405157a0d7c\\"] = function()
local ____exports = {}
____exports.value = true
return ____exports

end,
[\\"9124a2db32045398a4f097c567d1fe6fcb7ab900\\"] = function()
local ____exports = {}
do
    local ____foo = require(\\"5d05566c99dac259f6ff5742a268d405157a0d7c\\")
    local value = ____foo.value
    ____exports.value = value
end
return ____exports

end,
}
return require(\\"9124a2db32045398a4f097c567d1fe6fcb7ab900\\")"
`;

exports[`printer 1`] = `"-- Plugin"`;
